### PR TITLE
change checkbox state and raise change event with label/focus/selectionChange (closes #3348)

### LINKED ISSUE #3348: Button not clicked when .focus() called before
### What is your Test Scenario?
I'm trying to toggle a button. If my input-field (ace-editor) calls .focus() before, the button will be hovered but not clicked. If I do not have the .focus() call, the button is correctly clicked. It works fine without testcafe.
 
### What is the Current behavior?
Button is not toggled, when .focus() is called on the editor. It is toggled in any other case.
 
### What is the Expected behavior?
Click-Action from testcafe does not care about focus of components.
 
### What is your web application and your TestCafe test code?

 
Your website URL (or attach your complete example): https://jsfiddle.net/3yafrzwL/2/

<details>
<summary>Your complete test code (or attach your test files):</summary>

 
```ts
fixture('Button Bug').page('https://jsfiddle.net/3yafrzwL/2/');

test('click button in jsfiddle', async (t) => {
  const selector = Selector('#advancedWrapper');
  await t
    .switchToIframe(Selector('[name=result]'))
    .wait(3000)
    .click(selector)
    .expect(selector.find('.oj-button').hasClass('oj-selected'))
    .ok()
    .wait(3000);
});
```
</details>

<details>
<summary>Screenshots:</summary>

I get:

actual

I expect:
expected

```

```
</details>

### Steps to Reproduce:

1. Execute the provided code. It will open https://jsfiddle.net/3yafrzwL/2/
 
### Your Environment details:
 
* testcafe version:                  0.23.1
* node.js version:                   8.11.3
* command-line arguments:  node --no-deprecation --inspect-brk=49693 node_modules/testcafe/lib/cli/index.js chrome jsfiddle.ts -e --test "click button in jsfiddle"
* browser name and version: Chrome 71.0.3578
* platform and version:         Ubuntu 18.10
* other information: I'm using Oracle JET - Toggle Button (https://www.oracle.com/webfolder/technetwork/jet/jetCookbook.html?component=toggleButtons&demo=buttonCheckbox) and ace-editor. I cannot reproduce the problem without ace editor or with a normal button.

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: src/client/core/utils/dom.js [hunk 1]
```diff
@@ -51,6 +51,8 @@ export const findParent                             = hammerhead.utils.dom.findP
 export const getTopSameDomainWindow                 = hammerhead.utils.dom.getTopSameDomainWindow;
 
 export const isRadioButtonElement = el => isInputElement(el) && el.type === 'radio';
+export const isColorInputElement  = el => isInputElement(el) && el.type === 'color';
```
Comment (miherlosev) [inline src/client/core/utils/dom.js]: I don't know why it's necessary to reexport some methods from hammerhead.
Use the appropriate dom utils methods from [hammerhead](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/utils/dom.ts#L440) at once.

File: src/client/automation/utils/utils.js [hunk 2]
```diff
@@ -104,13 +104,20 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
...
 
-export function focusByRelatedElement (element) {
+export function getElementBoundToLabel (element) {
     const labelWithForAttr = domUtils.closest(element, 'label[for]');
 
     if (!labelWithForAttr)
-        return;
+        return null;
+
+    const boundElementId = labelWithForAttr.getAttribute('for');
```
Comment (miherlosev) [inline src/client/automation/utils/utils.js]: `return labelWithForAttr.control`

https://developer.mozilla.org/en-US/docs/Web/API/HTMLLabelElement/control

File: src/client/automation/playback/click/index.js [hunk 3]
```diff
@@ -195,34 +196,61 @@ export default class ClickAutomation extends VisibleElementAutomation {
     }
 
     _click (eventArgs) {
-        if (domUtils.isOptionElement(eventArgs.element))
+        const elementBoundToLabel = getElementBoundToLabel(eventArgs.element);
```
Comment (miherlosev) [inline src/client/automation/playback/click/index.js]: Can we split this complex method with such manner?
```js
if (isOptionElement) {}
else if (ColorInputElement) {}
else {....}
```

File: src/client/core/utils/dom.js (hunk 1 above)
Comment (AlexKamaev) [reply src/client/core/utils/dom.js]: `hammerhead` does not have such methods as `isRadioButtonElement`, `isColorInputElement` or `        isCheckboxElement`

Comment (miherlosev) [reply src/client/core/utils/dom.js]: Add them

Comment (AlexKamaev) [reply src/client/core/utils/dom.js]: sounds like a separate issue in hammerhead

**Comment by AndreyBelym** [comment]
It failed twice, I guess it's not a random failure. Likely Firefox dies with an exception when trying to show a native dialog.

```
[Regression](GH-1327)
    Typing in non-text input elements
[10:49:09] Error: This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (net.js:393:12)
    at write (/home/travis/build/testcafe-build-bot/testcafe/src/browser/provider/built-in/dedicated/firefox/marionette-client/index.js:65:26)
    at Generator.next (<anonymous>)
    at step (/home/travis/build/testcafe-build-bot/testcafe/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
    at /home/travis/build/testcafe-build-bot/testcafe/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
    at new Promise (<anonymous>)
    at new F (/home/travis/build/testcafe-build-bot/testcafe/node_modules/core-js/library/modules/_export.js:36:28)
    at /home/travis/build/testcafe-build-bot/testcafe/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
    at MarionetteClient._writeSocket (/home/travis/build/testcafe-build-bot/testcafe/src/browser/provider/built-in/dedicated/firefox/marionette-client/index.js:64:34)
    at _writeSocket (/home/travis/build/testcafe-build-bot/testcafe/src/browser/provider/built-in/dedicated/firefox/marionette-client/index.js:119:18)
```

File: src/client/automation/playback/click/click-command.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+export default function (eventState, eventArgs) {
+    const elementBoundToLabel = getElementBoundToLabel(eventArgs.element);
+    const isSelectElement     = domUtils.isSelectElement(eventArgs.element);
+    const isOptionElement     = domUtils.isOptionElement(eventArgs.element);
+    const isColorInputElement = domUtils.isColorInputElement(eventState.clickElement);
+    const isLabelledCheckbox  = elementBoundToLabel && domUtils.isCheckboxElement(elementBoundToLabel);
+
+    if (isSelectElement)
```
Comment (miherlosev) [inline src/client/automation/playback/click/click-command.js]: if (isSelectElement)...
else if (isOptionElement)...
else if (...)
else ...

File: src/client/automation/utils/utils.js [hunk 5]
```diff
@@ -104,13 +104,14 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
     });
 }
 
-export function focusByRelatedElement (element) {
+export function getElementBoundToLabel (element) {
     const labelWithForAttr = domUtils.closest(element, 'label[for]');
 
-    if (!labelWithForAttr)
-        return;
+    return labelWithForAttr ? labelWithForAttr.control : null;
```
Comment (miherlosev) [inline src/client/automation/utils/utils.js]: `return labelWithForAttr && labelWithForAttr.control`

File: src/client/core/utils/dom.js (hunk 1 above)
Comment (miherlosev) [reply src/client/core/utils/dom.js]: Today we will publish a new hammerhead version with appropriate dom utils.

File: src/client/automation/playback/click/click-command.js (hunk 4 above)
Comment (AlexKamaev) [reply src/client/automation/playback/click/click-command.js]: for what?

Comment (miherlosev) [reply src/client/automation/playback/click/click-command.js]: Sorry, I've missed the `return` statement. All is OK.

**Comment by AlexKamaev** [comment]
FPR

**Comment by AlexKamaev** [comment]
@miherlosev

**Pull request merged**